libraries/standard/mqtt/src/iot_mqtt_api.c

```diff
@@ -1054,126 +1035,124 @@ IotMqttError_t IotMqtt_Connect( const IotMqttNetworkInfo_t * pNetworkInfo,
-                                                pConnectInfo->keepAliveSeconds );
-
-    if( pNewMqttConnection == NULL )
-    {
-        status = IOT_MQTT_NO_MEMORY;
-        goto cleanup;
     }
     else
     {
-        /* Set the network connection associated with the MQTT connection. */
-        pNewMqttConnection->pNetworkConnection = pNetworkConnection;
-        pNewMqttConnection->ownNetworkConnection = ownNetworkConnection;
+        networkStatus = _createNetworkConnection( pNetworkInfo,
+                                                  &pNetworkConnection,
+                                                  &ownNetworkConnection );
 
-        /* Set the MQTT packet serializer overrides. */
-        #if IOT_MQTT_ENABLE_SERIALIZER_OVERRIDES == 1
-            pNewMqttConnection->pSerializer = pNetworkInfo->pMqttSerializer;
-        #else
-            pNewMqttConnection->pSerializer = NULL;
-        #endif
+        if( networkStatus != IOT_NETWORK_SUCCESS )
+        {
+            status = IOT_MQTT_NETWORK_ERROR;
+        }
     }
 
-    /* Set the MQTT receive callback. */
-    networkStatus = pNewMqttConnection->pNetworkInterface->setReceiveCallback( pNetworkConnection,
-                                                                               IotMqtt_ReceiveCallback,
-                                                                               pNewMqttConnection );
-
-    if( networkStatus != IOT_NETWORK_SUCCESS )
+    if( status == IOT_MQTT_SUCCESS )
     {
-        IotLogError( "Failed to set MQTT network receive callback." );
+        IotLogInfo( "Establishing new MQTT connection." );
 
-        status = IOT_MQTT_NETWORK_ERROR;
-        goto cleanup;
-    }
+        /* Initialize a new MQTT connection object. */
+        pNewMqttConnection = _createMqttConnection( pConnectInfo->awsIotMqttMode,
+                                                    pNetworkInfo,
+                                                    pConnectInfo->keepAliveSeconds );
 
-    /* Create a CONNECT operation. */
-    status = _IotMqtt_CreateOperation( pNewMqttConnection,
-                                       IOT_MQTT_FLAG_WAITABLE,
-                                       NULL,
-                                       &pOperation );
+        if( pNewMqttConnection == NULL )
+        {
+            status = IOT_MQTT_NO_MEMORY;
+        }
+        else
+        {
+            /* Set the network connection associated with the MQTT connection. */
+            pNewMqttConnection->pNetworkConnection = pNetworkConnection;
+            pNewMqttConnection->ownNetworkConnection = ownNetworkConnection;
+
+            /* Set the MQTT packet serializer overrides. */
+            #if IOT_MQTT_ENABLE_SERIALIZER_OVERRIDES == 1
+                pNewMqttConnection->pSerializer = pNetworkInfo->pMqttSerializer;
+            #else
+                pNewMqttConnection->pSerializer = NULL;
+            #endif
+        }
```

Comment:
The contents of this `else` block could also be at the beginning of the very next `if`. It makes no difference for this PR, but it would help in reducing complexity down the line due to a lower nesting level.